### PR TITLE
Release 21 07 debug

### PR DESCRIPTION
**Pull request by hhoegelo**
I found some out-of-bound array accesses and potential optimizations. The out-of-bounds triggered some segfaults in my artificial test cases, not sure if this correlate with the crashes found by our users.

### CODE REVIEW

File: projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h [hunk 1]
```diff
@@ -192,32 +194,30 @@ namespace Engine
       }
       inline Target_Param* globalChainFirst(const uint32_t _mcId)
       {
-        return &m_global.m_target[m_global.m_assignment.first(_mcId)];
-      }
-      inline bool globalChainRunning()
-      {
-        return m_global.m_assignment.running();
+        auto idx = m_global.m_assignment.first(_mcId);
+        return idx < 0 ? nullptr : &m_global.m_target[m_global.m_assignment.first(_mcId)];
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h:198]: I'd propose
`return idx < 0 ? nullptr : &m_global.m_target[idx)];`
as idx is already determined and valid

File: projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h [hunk 2]
```diff
@@ -192,32 +194,30 @@ namespace Engine
...
       }
+
       inline Target_Param* globalChainNext()
       {
-        return &m_global.m_target[m_global.m_assignment.next()];
+        auto idx = m_global.m_assignment.next();
+        return idx < 0 ? nullptr : &m_global.m_target[m_global.m_assignment.next()];
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h:204]: I'd propose
`return idx < 0 ? nullptr : &m_global.m_target[idx)];`
as idx is already determined and valid

File: projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h [hunk 3]
```diff
@@ -192,32 +194,30 @@ namespace Engine
...
+
       inline Target_Param* localChainFirst(const uint32_t _layer, const uint32_t _mcId)
       {
-        return &m_layer[_layer].m_target[m_layer[_layer].m_assignment.first(_mcId)];
-      }
-      inline bool localChainRunning(const uint32_t _layer)
-      {
-        return m_layer[_layer].m_assignment.running();
+        auto idx = m_layer[_layer].m_assignment.first(_mcId);
+        return idx < 0 ? nullptr : &m_layer[_layer].m_target[m_layer[_layer].m_assignment.first(_mcId)];
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h:210]: I'd propose
`return idx < 0 ? nullptr : &m_layer[_layer].m_target[idx];`
as idx is already determined and valid

File: projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h [hunk 4]
```diff
@@ -192,32 +194,30 @@ namespace Engine
...
       }
+
       inline Target_Param* localChainNext(const uint32_t _layer)
       {
-        return &m_layer[_layer].m_target[m_layer[_layer].m_assignment.next()];
+        auto idx = m_layer[_layer].m_assignment.next();
+        return idx < 0 ? nullptr : &m_layer[_layer].m_target[idx];
```
Comment (MatthiasSeeber) [inline projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h:216]: ... like this. Further, idx could/should be const in each case.

**Review by MatthiasSeeber — APPROVED** [review]
I just have some minor suggestions that optimize a bit further.

Nevertheless, the code is good to go and successfully tested (I sound checked several Presets and their modulation, they sound identical to current unstable).

File: projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h (hunk 1 above)
Comment (hhoegelo) [reply projects/epc/audio-engine/src/synth/c15-audio-engine/parameter_handle.h:198]: good catch, of course this shouldn't be evaluated twice - will fix

**Review by MatthiasSeeber — APPROVED** [review]

**Review by schlenkibus — APPROVED** [review]

**Pull request merged**